### PR TITLE
Allow support for nanosecond timeouts, and use nanoseconds consistent…

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
…ly throughout our code

### CODE REVIEW

**Comment by buchgr** [comment]
@carl-mastrangelo is there a specific reason for allowing nanos? I wouldn't expect `ScheduledThreadPoolExecutor` to be anywhere close to nanosecond accuracy?

**Comment by carl-mastrangelo** [comment]
@buchgr It's part of the spec: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

**Comment by buchgr** [comment]
@carl-mastrangelo ahh okay. makes sense then :).

PS: Just if you are curios https://gist.github.com/buchgr/18eb11e6eb228b316b66

Gives me (in nanos, on else idle system):
```
Jakobs-MBP-2:src buchgr$ java -Xms2g -Xmx2g -verbose:gc Main
max: 195730, min: 760, avg: 22282, median: 15913, runs: 10000
Jakobs-MBP-2:src buchgr$ java -Xms2g -Xmx2g -verbose:gc Main
max: 128789, min: 657, avg: 16498, median: 11246, runs: 10000
Jakobs-MBP-2:src buchgr$ java -Xms2g -Xmx2g -verbose:gc Main
max: 150990, min: 651, avg: 21597, median: 16592, runs: 10000
```

**Comment by carl-mastrangelo** [comment]
Also, I made an error earlier this year where I didn't convert micros to nanos.  Since we don't have a proper Duration object, I am making all durations nanoseconds to avoid possible errors.

I just happened to notice we weren't spec compliant when I looked through this.  

Fun fact: the TimeUnit conversions saturate the long rather than overflow when the value is too big.  The maximum allowed timeout is 99,999,999H which overflows a long.

File: core/src/main/java/io/grpc/internal/ClientTransport.java [hunk 1]
```diff
@@ -136,12 +136,12 @@ ClientStream newStream(MethodDescriptor<?, ?> method,
...
      * Invoked when a ping is acknowledged. The given argument is the round-trip time of the ping,
-     * in microseconds.
+     * in nanoseconds.
      *
-     * @param roundTripTimeMicros the round-trip duration between the ping being sent and the
+     * @param roundTripTimeNanos the round-trip duration between the ping being sent and the
      *     acknowledgement received
      */
-    void pingAcknowledged(long roundTripTimeMicros);
+    void pingAcknowledged(long roundTripTimeNanos);
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: Although this is internal API, there may be people using it. This would give them wrong data and it's almost impossible for them to notice it. Please keep the time as micros.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: This is one of the few places that we have allowed ourselves backwards incompatibility, and I think we should be able to exercise it.  People who depend on internal implementation details before GA have implicitly accepted that behavior is not stable.  This change makes our API completely uniform and I would argue is necessary before we publicly commit to never changing it.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: Backwards incompatibility means we are free to break the API, but such silent semantic change is unsettling to me.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: I could change the method name if you'd like.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: SGTM.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: Done.

**Comment by zhangkun83** [comment]
LGTM

File: core/src/main/java/io/grpc/internal/GrpcUtil.java [hunk 2]
```diff
@@ -39,6 +39,8 @@
 import com.google.common.base.Joiner;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Splitter;
+import com.google.common.collect.BiMap;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: @madongfly, should we use BiMap in core/? Would that disappoint some of our Android users?

File: core/src/main/java/io/grpc/internal/ClientTransport.java [hunk 3]
```diff
@@ -136,18 +136,18 @@ ClientStream newStream(MethodDescriptor<?, ?> method,
...
      * Invoked when a ping is acknowledged. The given argument is the round-trip time of the ping,
-     * in microseconds.
+     * in nanoseconds.
      *
-     * @param roundTripTimeMicros the round-trip duration between the ping being sent and the
+     * @param roundTripTimeNanos the round-trip duration between the ping being sent and the
      *     acknowledgement received
      */
-    void pingAcknowledged(long roundTripTimeMicros);
+    void onSuccess(long roundTripTimeNanos);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: Please don't mix arbitrary API renames with unrelated changes.

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ClientTransport.java]: Oh. I guess this was detect that those APIs needed to convert from micros to nanos. I guess that's fine, although I would have probably preferred to leave this API as-is since it is primarily there for external use (we don't use it at all internally).

File: core/src/main/java/io/grpc/internal/GrpcUtil.java [hunk 4]
```diff
@@ -431,54 +436,41 @@ public void close(ScheduledExecutorService instance) {
    */
   @VisibleForTesting
   static class TimeoutMarshaller implements Metadata.AsciiMarshaller<Long> {
+    private static final BiMap<TimeUnit, Character> UNITS =
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: This should be `ImmutableBiMap` instead of `BiMap` with a comment that it depends on the `ImmutableBiMap` iteration order.

**Comment by ejona86** [comment]
@carl-mastrangelo, although the deadline supports nanosecond precision, implementations are not required to support nanosecond accuracy. Rounding up to the nearest microsecond or millisecond is permitted if that implementation wants to do so. So this change wasn't required from any specification perspective.

That said, I'm all for using nanos for durations in general. And dealing saturating instead of wrapping on overflowing is nice.

File: core/src/main/java/io/grpc/internal/GrpcUtil.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: Is it not well known that the immutable classes have consistent iteration order?

**Comment by carl-mastrangelo** [comment]
@ejona86 This change was not about precision, but rather about removing the possibility of conversion bugs.  (despite the title of this PR)

File: core/src/main/java/io/grpc/internal/GrpcUtil.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: FYI this could be changed to a regular map if needed.

Comment (madongfly) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: We used to remove the usages of guava collect package, because GMS core doesn't like it, I'm not very sure about the current status, but I'd assume they still don't like it.

File: core/src/main/java/io/grpc/internal/GrpcUtil.java (hunk 4 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: It isn't well-known that you are depending on the iteration order. Also, you are depending on the immutable implementation, not the `BiMap` interface. If I changed the construction to `HashBiMap` I would be surprised that it would break, as typically it would be considered mostly interchangeable.

**Comment by ejona86** [comment]
@carl-mastrangelo, ah, sorry. I misread @buchgr's comment, and thus mis-understood your response that linked to the spec.

File: core/src/main/java/io/grpc/internal/GrpcUtil.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/GrpcUtil.java]: I'm personally in the camp of less is more, but I created #1298 to change this.
